File: algorithms/sorting/selection_sort/selection_sort.cpp
#include <iostream>

using namespace std;

#define REP(i, n) for(int i = 0; i < n; i++)

void selection_sort(int *ar, int size)
{
  int smallest, pos, j;
  REP(i, size - 1) {
    
    smallest = ar[i];
    pos = i;

    for(j = i + 1; j < size; j++) {
      if (smallest > ar[j]) {
        smallest = ar[j];
        pos = j;
      }
    }

    if (pos != i) {
      ar[pos] = ar[i];
      ar[i] = smallest;
    }
  }
}

void display_array(int *ar, int size)
{
  REP(i, size) {
    cout << ar[i] << "  ";
  }
}

int main()
{
  int *ar;
  int size;

  cout << "What size arrray do you want?";
  cin >> size;

  ar = new int[size];
  cout << "Enter the array elemnets : ";
  REP(i, size)
    cin >> ar[i];

  cout << "The current array is : " << endl;
  display_array(ar, size);

  selection_sort(ar, size);

  cout << endl << "The sorted array is : " << endl;
  display_array(ar, size);
}

File: algorithms/sorting/selection_sort/selection_sort.rb
class Array
  def selection_sort!
    size.times do |i|
      smallest = self[i]
      pos = i
      
      ((i + 1)...self.size).each do |j|
        if smallest > self[j]
          smallest = self[j]
          pos = j
        end
      end

      if pos != i
        self[pos] = self[i]
        self[i] = smallest
      end
    end
  end
end

def interact
  puts "What size array do you want?"
  size = gets.chomp.to_i

  array = []
  puts "Enter the array elements : "
  size.times { array.push(gets.chomp.to_i) }

  puts "Current array is : "
  puts array.inspect

  array.selection_sort!

  puts "Sorted array is : "
  puts array.inspect
end

interact